File: headless/src/main/org/nlogo/prim/_patchhereinternal.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.Patch;
import org.nlogo.agent.Turtle;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;

// needed by _patchat.optimize() because regular _patchhere is turtle-only

public final strictfp class _patchhereinternal
    extends Reporter {
  @Override
  public Syntax syntax() {
    return Syntax.reporterSyntax
        (Syntax.PatchType(), "-TP-");
  }

  @Override
  public Object report(Context context) {
    return report_1(context);
  }

  public Patch report_1(Context context) {
    if (context.agent instanceof Patch) {
      return (Patch) context.agent;
    } else if (context.agent instanceof Turtle) {
      return ((Turtle) context.agent).getPatchHere();
    } else {
      return world.fastGetPatchAt(0, 0);
    }
  }
}

File: src/main/org/nlogo/prim/_patchhereinternal.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.agent.{ Turtle, Patch }
import org.nlogo.api.Syntax
import org.nlogo.nvm.{ Reporter, Context }

// needed by _patchat.optimize() because regular _patchhere is turtle-only

class _patchhereinternal extends Reporter {

  override def syntax =
    Syntax.reporterSyntax(
      Syntax.PatchType, "-TP-")

  override def report(context: Context): Patch =
    report_1(context)

  def report_1(context: Context): Patch =
    context.agent match {
      case p: Patch => p
      case t: Turtle => t.getPatchHere
    }

}
